doc/GUIDE.md

```diff
@@ -0,0 +1,822 @@
+```
+
+Here, we could run `jextract` and filter out `A`, like so:
+
+```sh
+$ jextract --include-var aVar test.h
+```
+
+However, doing so would lead to broken generated code, as the layout of the global variable
+`aVar` depends on the layout of the excluded struct `A`.
+
+In such cases, `jextract` will report the missing dependency and terminate without
+generating any bindings:
+
+```txt
+ERROR: aVar depends on A which has been excluded
+```
+
+## Tracing
+
+It is sometimes useful to inspect the parameters passed to a native call, especially when
+diagnosing application bugs and/or crashes. The code generated by the `jextract` tool
+supports _tracing_ of native calls, that is, parameters passed to native calls can be
+printed on the standard output.
+
+To enable the tracing support, just pass the `-Djextract.trace.downcalls=true` flag to the
+launcher used to start the application. Below we show an excerpt of the output when
+running the [OpenGL example](samples/opengl) with tracing support enabled:
+
+```txt
+glutInit(MemorySegment{ address: 0x7fa6b03d6400, byteSize: 4 }, MemorySegment{ address: 0x7fa6b03d6400, byteSize: 4 })
+glutInitDisplayMode(18)
+glutInitWindowSize(900, 900)
+glutCreateWindow(MemorySegment{ address: 0x7fa6b03f8e70, byteSize: 14 })
+glClearColor(0.0, 0.0, 0.0, 0.0)
+glShadeModel(7425)
+glLightfv(16384, 4611, MemorySegment{ address: 0x7fa6b03de8d0, byteSize: 16 })
+glLightfv(16384, 4608, MemorySegment{ address: 0x7fa6b0634840, byteSize: 16 })
+glLightfv(16384, 4609, MemorySegment{ address: 0x7fa6b0634840, byteSize: 16 })
+glLightfv(16384, 4610, MemorySegment{ address: 0x7fa6b0634840, byteSize: 16 })
+glMaterialfv(1028, 5633, MemorySegment{ address: 0x7fa6b0634860, byteSize: 4 })
+glEnable(2896)
+glEnable(16384)
+glEnable(2929)
+glutDisplayFunc(MemorySegment{ address: 0x7fa6a002e820, byteSize: 0 })
+glutIdleFunc(MemorySegment{ address: 0x7fa6a015a620, byteSize: 0 })
+glutMainLoop()
+glClear(16640)
+glPushMatrix()
+glRotatef(-20.0, 1.0, 1.0, 0.0)
+glRotatef(0.0, 0.0, 1.0, 0.0)
+glutSolidTeapot(0.5)
+```
+
+## Preprocessor Definitions
+
+C header files are processed by a pre-processor by a compiler before they are inspected
+further. It is possible for a header file to contain so-called 'compiler switches', which
+can be used to conditionally generate code based on the value of a macro, for instance:
+
+```c
+#ifdef MY_MACRO
+int x = 42;
+#else
+int x = 0;
+#endif
+```
+
+The value of these macros also affects the behavior of jextract. Therefore, jextract
+supports setting macro values on the command line using the `-D` or
+`--define-macro <macro>=<value>` option. For instance, we can use `-D MY_MACRO` to set
+the value of `MY_MACRO` in the above snippet to `1`, and trigger the first 'branch' of the
+compiler switch, thereby defining `int x = 42`.
+
+Please note that other header files included by jextract may also define macro values using
+the `#define` pre-processor directive. Therefore it is often important in which order
```

Comment:
There seems to be issues with this sentence. Possible suggestion:

`It is therefore important to notice the order in which header files are processed by a compiler, as feeding header files to jextract in the wrong order may result in weird errors due to missing macro definitions`